### ISSUE TITLE
`text/event-stream` not supported

### ISSUE THREAD

**Issue by rome-user** [ogen-go/ogen#1375]
### What version of ogen are you using?

```bash
$ go list -m github.com/ogen-go/ogen
github.com/ogen-go/ogen v1.8.1
```

### Can this issue be reproduced with the latest version?

Yes

### What did you do?

Consider the following schema. Save it as `openapi.yaml`.

```yaml
openapi: 3.0.0
info:
  title: Example API
  version: "1.0"
paths:
  /events:
    get:
      summary: SSE example
      responses:
        "200":
          description: OK
          content:
            text/event-stream:
              schema:
                type: string
```

Run the following terminal commands

```
$ go mod init repro
$ go get github.com/ogen-go/ogen/cmd/ogen
$ go run github.com/ogen-go/ogen/cmd/ogen --target gen/api -package api --clean openapi.yaml
$ go mod tidy
```

Create a `main.go` file with the following contents.

<details>
<summary>Click to view server implementation</summary>

```go
package main

import (
	"context"
	"io"
	"log"
	"net/http"
	"time"

	"repro/api"
)

func main() {
	srv, err := api.NewServer(&service{})
	if err != nil {
		log.Fatal(err)
	}

	if err := http.ListenAndServe(":8080", srv); err != nil {
		log.Fatal(err)
	}
}

type service struct{}

func (s *service) EventsGet(ctx context.Context) (api.EventsGetOK, error) {
	r, w := io.Pipe()

	go func() {
		defer w.Close()
		for ctx.Err() == nil {
			select {
			case <-ctx.Done():
				return
			case <-time.After(1 * time.Second):
				w.Write([]byte("data: ping\n\n"))
			}
		}
	}()

	return api.EventsGetOK{Data: r}, nil
}
```
</details>

Finally, start the server.

```
$ go run main.go
```

### What did you expect to see?

Client can stream the response. e.g.

```
$ curl -N http://localhost:8080/events
data: ping
data: ping
...
```

### What did you see instead?

The generated code in `oas_response_encoders_gen.go` looks like the following.

<details>
<summary>Click to view generated code</summary>

```go
func encodeEventsGetResponse(response EventsGetOK, w http.ResponseWriter, span trace.Span) error {
	w.Header().Set("Content-Type", "text/event-stream")
	w.WriteHeader(200)
	span.SetStatus(codes.Ok, http.StatusText(200))

	writer := w
	if _, err := io.Copy(writer, response); err != nil {
		return errors.Wrap(err, "write")
	}

	return nil
}
```
</details>

The call to `io.Copy` prevents streaming of response. This results in clients not receiving data.

**Comment by rokf** [comment]
I'm personally interested in this functionality so that I could generate boilerplate handler code for https://data-star.dev with ogen.

What I do currently is that I have two OpenAPI specifications for an application - one API that returns HTML for the UI and one that returns JSON for M2M communication.

With support for `text/event-stream` I could squeeze in the SSE endpoints for Datastar into my ogen HTML API - currently I need a custom set of handlers next to those managed by ogen and combine them all with a mux.

**Comment by rokf** [comment]
As a workaround you can use middleware to pass and retrieve the raw request and response objects - https://github.com/ogen-go/ogen/discussions/1252.

I've tested this approach with https://github.com/starfederation/datastar/tree/main/sdk/go and it works fine.

The only issue I found was that Ogen writes a an error log saying that there's an superfluous header write once the stream is finished - it doesn't seem to break anything though.

**Comment by rokf** [comment]
I made another discovery today, which makes the current workaround totally usable as it seems - the superfluous header error is gone 😄

You'll need at least one dummy response in the OpenAPI specification:

```
responses:
        "default":
          $ref: "#/components/responses/default"
```

Then you can use a dummy no-op error and watch for it in Ogen's error handler, i.e.:

```go
var Noop = errors.New("noop-error")

func ErrorHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, err error) {
	if errors.Is(err, Noop) {
		return
	}
```

In the Ogen HTTP handlers where manually writing your responses you can then return the no-op error at the end of the SSE "session" like:

```go
        r := util.RequestFromContext(ctx)
	w := util.ResponseFromContext(ctx)

        // ...

	sse := datastar.NewSSE(w, r)
	err = sse.MergeFragments(buff.String())
	if err != nil {
		return nil, err
	}

	return nil, herrors.Noop
}
```

**Comment by rokf** [comment]
@joeblew999 the repo in which we're using patterns described above is private. It's for a project at work. I don't have a public repo with this code.

Regarding Datastar and its use together with ogen - you should be able to find all the information you'll need to make it work in this thread as long as you're familiar with the basics of ogen.

For making a HTTP server that combines two or more HTTP servers generated by ogen you can use `http.NewServeMux` and its handle method to attach server instances under different prefixes. You might have to strip prefixes with `http.StripPrefix` - it depends on your endpoints in the OpenAPI specifications.

**Comment by barneyferry** [comment]
I noticed a similar issue on the generated client code, because ogen (like most openapi codegen tools) waits for the entire response in order to validate it, you don't get a live streaming response in `Data io.Reader` due to:

```go
		switch {
		case ct == "text/event-stream":
			reader := resp.Body
			b, err := io.ReadAll(reader)
			if err != nil {
				return res, err
			}
```

Now, I'd be happy with an escape-hatch here as it seems there's [not yet a consensus on what the OpenAPI spec should do with a `text/event-stream`](https://github.com/OAI/OpenAPI-Specification/discussions/4171) content type in terms of validation and it may be out of scope for ogen.

Perhaps an `x-ignore` for particular responses, so you still get the generated server/client functions with typed params but they simply return a bare, untouched `http.Response`.

**Comment by barneyferry** [comment]
Yeah, or alternatively expose generated path builder functions so users can easily construct a custom http.Request{} + Do() chain, this way you get the benefits of the generated code while being composable with stdlib without needing to add edge-case feature requests.

**Comment by zguydev** [comment]
SSE client generation was added in the latest release, you can read here: https://github.com/ogen-go/ogen#sse

I plan on implementing the SSE server generation feature in the near future. As a fallback, `x-ogen-raw-response` can be used for now. Note that `x-ogen-raw-response` takes priority over SSE client generation
